### PR TITLE
Owner check for contents of data directory, not directory itself

### PR DESCRIPTION
**Pull request by bkuschel**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- change startup check for owner of default clickhouse data directory to check contents of directory instead (#36003)

### CODE REVIEW

File: programs/keeper/Keeper.cpp [hunk 1]
```diff
@@ -288,20 +288,26 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
...
+    const auto effective_user = getUserName(effective_user_id);
+    fs::directory_iterator dir_end;
+    for (fs::directory_iterator dir_it(path); dir_it != dir_end; ++dir_it)
     {
-        const auto effective_user = getUserName(effective_user_id);
-        const auto data_owner = getUserName(statbuf.st_uid);
-        std::string message = "Effective user of the process (" + effective_user +
-            ") does not match the owner of the data (" + data_owner + ").";
-        if (effective_user_id == 0)
+        const fs::path kpfile = dir_it->path();
```
Comment (alexey-milovidov) [inline programs/keeper/Keeper.cpp:295]: It's not obvious, what is `k` and `p` in `kpfile`.

File: programs/keeper/Keeper.cpp [hunk 2]
```diff
@@ -288,20 +288,26 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
...
+        if (kpfile.filename().compare("lost+found") != 0 && stat(kpfile.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
```
Comment (alexey-milovidov) [inline programs/keeper/Keeper.cpp:296]: You can write `==` for the same effect.

File: programs/keeper/Keeper.cpp [hunk 3]
```diff
@@ -288,20 +288,26 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
...
         {
-            message += " Run under 'sudo -u " + data_owner + "'.";
-            throw Exception(message, ErrorCodes::MISMATCHING_USERS_FOR_PROCESS_AND_DATA);
-        }
-        else
-        {
-            LOG_WARNING(log, fmt::runtime(message));
+            const auto data_owner = getUserName(statbuf.st_uid);
+            std::string message = "Effective user of the process (" + effective_user +
```
Comment (alexey-milovidov) [inline programs/keeper/Keeper.cpp:299]: Let's replace with `fmt::format` just in case.

File: programs/keeper/Keeper.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply programs/keeper/Keeper.cpp:295]: I guess it is `keeper`. Then the variable should be named `keeper_file`.

File: programs/server/Server.cpp [hunk 4]
```diff
@@ -820,20 +820,26 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
+            std::string message = "Effective user of the process (" + effective_user +
+                ") does not match the owner (" + data_owner + ") of a data file.";
+            if (effective_user_id == 0)
+            {
+                message += " Run under 'sudo -u " + data_owner + "'.";
+                throw Exception(message, ErrorCodes::MISMATCHING_USERS_FOR_PROCESS_AND_DATA);
+            }
+            else
+            {
+                global_context->addWarningMessage(message);
```
Comment (alexey-milovidov) [inline programs/server/Server.cpp:840]: Let's make sure only one warning is added if multiple files found.
Reason: many warnings about the same issue will be too annoying.

File: programs/server/Server.cpp [hunk 5]
```diff
@@ -820,20 +820,25 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
+    fs::directory_iterator dir_end;
+    for (fs::directory_iterator dir_it(path); dir_it != dir_end; ++dir_it)
     {
-        const auto effective_user = getUserName(effective_user_id);
-        const auto data_owner = getUserName(statbuf.st_uid);
-        std::string message = "Effective user of the process (" + effective_user +
-            ") does not match the owner of the data (" + data_owner + ").";
-        if (effective_user_id == 0)
+        const fs::path dbfile = dir_it->path();
+        if (dbfile.filename() == "lost+found" && stat(dbfile.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
```
Comment (azat) [inline programs/server/Server.cpp:828]: Seems that here is a typo?

```suggestion
        if (dbfile.filename() != "lost+found" && stat(dbfile.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
```

File: programs/keeper/Keeper.cpp [hunk 6]
```diff
@@ -288,20 +288,25 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
...
+    fs::directory_iterator dir_end;
+    for (fs::directory_iterator dir_it(path); dir_it != dir_end; ++dir_it)
     {
-        const auto effective_user = getUserName(effective_user_id);
-        const auto data_owner = getUserName(statbuf.st_uid);
-        std::string message = "Effective user of the process (" + effective_user +
-            ") does not match the owner of the data (" + data_owner + ").";
-        if (effective_user_id == 0)
+        const fs::path keeper_file = dir_it->path();
+        if (keeper_file.filename() == "lost+found" && stat(keeper_file.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
```
Comment (azat) [inline programs/keeper/Keeper.cpp:296]: Same here

```suggestion
        if (keeper_file.filename() != "lost+found" && stat(keeper_file.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
```

File: programs/keeper/Keeper.cpp [hunk 7]
```diff
@@ -288,20 +288,25 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
...
         {
-            message += " Run under 'sudo -u " + data_owner + "'.";
-            throw Exception(message, ErrorCodes::MISMATCHING_USERS_FOR_PROCESS_AND_DATA);
-        }
-        else
-        {
-            LOG_WARNING(log, fmt::runtime(message));
+            const auto data_owner = getUserName(statbuf.st_uid);
+            std::string message = fmt::format("Effective user of the process ({}) does not match the owner ({}) of a data file.",effective_user,data_owner);
```
Comment (azat) [inline programs/keeper/Keeper.cpp:299]:
```suggestion
            std::string message = fmt::format("Effective user of the process ({}) does not match the owner ({}) of a data file.", effective_user, data_owner);
```

File: programs/server/Server.cpp [hunk 8]
```diff
@@ -820,20 +820,25 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
         {
-            message += " Run under 'sudo -u " + data_owner + "'.";
-            throw Exception(message, ErrorCodes::MISMATCHING_USERS_FOR_PROCESS_AND_DATA);
-        }
-        else
-        {
-            global_context->addWarningMessage(message);
+            const auto data_owner = getUserName(statbuf.st_uid);
+            std::string message = fmt::format("Effective user of the process ({}) does not match the owner ({}) of a data file.",effective_user,data_owner);
```
Comment (azat) [inline programs/server/Server.cpp:831]:
```suggestion
            std::string message = fmt::format("Effective user of the process ({}) does not match the owner ({}) of a data file.", effective_user, data_owner);
```

File: programs/keeper/Keeper.cpp [hunk 9]
```diff
@@ -288,20 +288,25 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
     /// Check that the process user id matches the owner of the data.
     const auto effective_user_id = geteuid();
     struct stat statbuf;
-    if (stat(path.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
+    const auto effective_user = getUserName(effective_user_id);
```
Comment (azat) [inline programs/keeper/Keeper.cpp:291]: By the way you may not have any mapping for uid to login, so I guess it is better to uid into error message too.

Comment (bkuschel) [reply programs/keeper/Keeper.cpp:291]: yes, we have this situation. It seems to use uid for effective_user if there is no username.

File: programs/keeper/Keeper.cpp (hunk 7 above)
Comment (bkuschel) [reply programs/keeper/Keeper.cpp:299]: Yes, thanks

**Comment by bkuschel** [comment]
Smoke test is failing because of file permission problem in the docker image. Checking.

**Comment by alexey-milovidov** [comment]
See the errors from ShellCheck.

File: docker/test/split_build_smoke_test/run.sh [hunk 10]
```diff
@@ -4,6 +4,7 @@ set -x
 
 install_and_run_server() {
     mkdir /unpacked
+    cd /unpacked
```
Comment (alexey-milovidov) [inline docker/test/split_build_smoke_test/run.sh:7]: Why this change needed?

Comment (bkuschel) [reply docker/test/split_build_smoke_test/run.sh:7]: My hunch is that changing the working directory to unpacked my avoid some of the volume mounts off of root directories that could be owned by ubuntu.

Comment (azat) [reply docker/test/split_build_smoke_test/run.sh:7]: >Actually, the fix is not good here either

Why not? LGTM.

>I'm not 100% sure what user 1000/ubuntu is owning here.

```sh
$ docker run --rm -it clickhouse/split-build-smoke-test find / -maxdepth 1 \! -user 0
/sys
/proc
```

Obviously.

BTW You may improve logging to show it

This change is required anyway to avoid clickhouse check every directory in the `/`, you can use `clickhouse-server -- --path /unpacked` by the way (plus if you `cd` here, then `tar` does not requires `-C` option)

Comment (bkuschel) [reply docker/test/split_build_smoke_test/run.sh:7]: Yes, that's what I suspected but couldn't reproduce it in my default docker setup everything mounts as root. Thank you for the confirmation! I'll definitely try your path suggestion.

File: tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh [hunk 11]
```diff
@@ -5,9 +5,12 @@ CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 # shellcheck source=../shell_config.sh
 . "$CUR_DIR"/../shell_config.sh
 
+mkdir $CUR_DIR/01737_clickhouse
```
Comment (azat) [inline tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh:8]:
```suggestion
mkdir $CUR_DIR/01737_clickhouse
function cleanup() { rm -fr $CUR_DIR/01737_clickhouse; }
trap cleanup EXIT
```

*Untested code*

File: tests/queries/0_stateless/01594_too_low_memory_limits.sh [hunk 12]
```diff
@@ -19,9 +19,12 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 # shellcheck source=../shell_config.sh
 . "$CURDIR"/../shell_config.sh
 
+mkdir $CURDIR/01594_clickhouse
```
Comment (azat) [inline tests/queries/0_stateless/01594_too_low_memory_limits.sh:22]: Same here, it is better to remove the folder, after the test

**Comment by alexey-milovidov** [comment]
Need to merge with master after #36337.

**Comment by bkuschel** [comment]
@Mergifyio update

**Comment by bkuschel** [comment]
@azat @alexey-milovidov is this good to go? Failures seem unrelated.

One thing, this may break existing deployments where the contents of data directory contains files not owned by clickhouse user. Should this be added into migration notes?

**Comment by alexey-milovidov** [comment]
It is waiting code review.

**Comment by alexey-milovidov** [comment]
> Performance Comparison (actions) [1/4] — Failed to parse the report. 

This look suspicious for me.

**Comment by bkuschel** [comment]
Performance tests fixed now.

File: docker/test/performance-comparison/compare.sh [hunk 13]
```diff
@@ -267,7 +270,7 @@ function run_tests
         # Check that both servers are alive, and restart them if they die.
         clickhouse-client --port $LEFT_SERVER_PORT --query "select 1 format Null" \
             || { echo $test_name >> left-server-died.log ; restart ; }
-        clickhouse-client --port $RIGHT_SERVER_PORT --query "select 1 format Null" \
+        clickhouse-client --port $RIGHT_SERVER_PORT--query "select 1 format Null" \
```
Comment (azat) [inline docker/test/performance-comparison/compare.sh:273]: Hm, this change looks wrong, and by the way performance tests are broken.

**Comment by azat** [comment]
>Performance tests fixed now.

Not really:

>Performance Comparison (actions) [1/4] — **Errors while building the report.**

File: programs/keeper/Keeper.cpp (hunk 9 above)
Comment (azat) [reply programs/keeper/Keeper.cpp:291]: I still don't see UIDs in the error message.

**Comment by azat** [comment]
@bkuschel see also https://github.com/ClickHouse/ClickHouse/pull/36073#discussion_r862823733 (I can't mark it as unresolved)

File: programs/keeper/Keeper.cpp (hunk 9 above)
Comment (bkuschel) [reply programs/keeper/Keeper.cpp:291]: What do you see, is it blank? 
For example, this is what was printing when some of the tests were failing:
```
Effective user of the process (root) does not match the owner (1000) of a data 
```
In this case I believe user 1000 (`ubuntu` in the host) is not in passwd and prints `1000`. This is what the old implementation did as well.  As this code was there originally, so you should been seeing it with the original implementation as well?

Comment (bkuschel) [reply programs/keeper/Keeper.cpp:291]: If I understand correctly `getpwuid_r` doesn't return an error if the password entry doesn't exist but rather returns no `result`. The code seems to be working on that basis...

```cpp
    const auto error = getpwuid_r(user_id, &passwd_entry, buffer.data(), buffer_size, &result);

    if (error)
        throwFromErrno("Failed to find user name for " + toString(user_id), ErrorCodes::FAILED_TO_GETPWUID, error);
    else if (result)
        return result->pw_name;
    return toString(user_id);
````

Comment (azat) [reply programs/keeper/Keeper.cpp:291]: Yes, you are right, I haven't looked at `getUserName()`, thanks.

File: tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh [hunk 14]
```diff
@@ -26,7 +29,8 @@ function cleanup()
     echo "Test failed. Server log:"
     cat clickhouse-server.log
     rm -f clickhouse-server.log
-
+    rm -fr $CUR_DIR/01737_clickhouse
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh:32]: This looks extremely dangerous.
What if `$CUR_DIR` contains whitespaces?

**Comment by alexey-milovidov** [comment]
Cannot edit your files:

Screenshot_20220507_145155

File: tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh (hunk 14 above)
Comment (bkuschel) [reply tests/queries/0_stateless/01737_clickhouse_server_wait_server_pool_long.sh:32]: i'll fix

**Comment by bkuschel** [comment]
@alexey-milovidov last failures seem unrelated (minio?) as far as i can tell. Is this good to go?

**Comment by bkuschel** [comment]
Some stateful tests need fix.

**Comment by bkuschel** [comment]
Update on this..  trying to run the github actions directly in a fork to work through the infrastructure issues better.

**Comment by alexey-milovidov** [comment]
The test still doesn't work.

**Comment by alexey-milovidov** [comment]
Please reopen or resubmit when ready.
